### ISSUE TITLE
[bug] Inconsistent Logout Flow

### ISSUE THREAD

**Issue by theBeginner86** [meshery/meshery#9129]
### Current Behavior

When a user logs out of Meshery then they experience an odd behavior where they see empty state in Meshery UI.
In the process of logging out, even through the user session is cleared and specific `token` named cookie is removed the local storage, still they are not redirected to login screen.
The above behavior is usually observed in playground environment. For the case of playground env, CORS errors are also observed in the console/network tab of the browser.

### Expected Behavior

Once users logs out then they should be redirected to login screen.

### Environment

- **Host OS:** Mac Linux Windows
- **Platform:** Docker or Kubernetes
- **Meshery Server Version:** stable-v
- **Meshery Client Version:** stable-v

---
### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/contributing/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and [designs for Meshery UI](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI) in Figma [(open invite)](https://www.figma.com/team_invite/redeem/qJy1c95qirjgWQODApilR9)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](http://discuss.meshery.io) and [Community Slack](https://slack.meshery.io)

**Comment by theBeginner86** [comment]
@aabidsofi19 @Yashsharma1911 please consider adding your findings here, if I missed documenting them.

**Comment by Shiva953** [comment]
Can I take up this Issue?

**Comment by Shiva953** [comment]
I'm not sure tho, but I suspect that this is the part which might be causing the time delay :

https://github.com/meshery/meshery/blob/008a6e03e72485f793f6344a5d3eac507905a3e9/provider-ui/pages/auth/login.js#L22C1-L43C4

However, the CORS Error source is still unclear to me.

**Comment by leecalcote** [comment]
Thanks for jumping here, @Shiva953. 👍 Please keep your notes coming...

**Comment by coder12git** [comment]
@theBeginner86 this is the lpart of code that's causing this bug - https://github.com/meshery/meshery/blob/008a6e03e72485f793f6344a5d3eac507905a3e9/provider-ui/components/Provider.js#L77-L98

Actually we are making cross-origin request with `credentials: include`,  but header in the response is wildcard , to fix this we need to specify the exact origin `https://playground.meshery.io/` in the header of the response. This is something which will fix that CORS error, and probably that bug too.

If the proposed solution seems correct to you, then I would like to work on this issue :)

**Comment by Shiva953** [comment]
Seems reasonable @coder12git

**Comment by coder12git** [comment]
> Seems reasonable @coder12git

Let's see what @theBeginner86 says

**Comment by theBeginner86** [comment]
> @theBeginner86 this is the part of code that's causing this bug -
> 
> https://github.com/meshery/meshery/blob/008a6e03e72485f793f6344a5d3eac507905a3e9/provider-ui/components/Provider.js#L77-L98
> 
> Actually we are making cross-origin request with `credentials: include`, but header in the response is wildcard , to fix this we need to specify the exact origin `https://playground.meshery.io/` in the header of the response. This is something which will fix that CORS error, and probably that bug too.
> 
> If the proposed solution seems correct to you, then I would like to work on this issue :)

@coder12git
This is a reasonable argument. Thanks for bringing this up.
Though that wouldn't solve the current issue at hand. Because for the case of playground deployment, we never land on Provider's page (as we have custom [build](https://github.com/meshery/meshery/blob/master/install/playground/docker/docker-compose.yaml#L10) that selects Meshery provider through an env variable that skips the Provider UI where users are given choice to select the providers). You can confirm this by checking the network tab while you navigating to Playground.

There is a quick way to reproduce this issue locally by running this make target: https://github.com/meshery/meshery/blob/master/Makefile#L184
(To test your local changes)


The CORS issue is due to the fact that Meshery Server is trying to directly reach out to one of the Identity Servers of Remote Provider and it rejects the requests because it has been configured to only listen to the requests that originates from https://meshery.layer5.io.
![Screenshot 2023-10-21 at 4 25 26 PM](https://github.com/meshery/meshery/assets/73700530/fc9d6b8b-02ee-4795-889e-71d37b9b1f23)

**Comment by coder12git** [comment]
@Shiva953 are you working on this bug? Let me know, because i also want to work on it, in case if you are not working :)

**Comment by Shiva953** [comment]
@coder12git I'm working on it since yesterday, but im still unable to relate the CORS Error source with the delay in having the login page, and the stateless screen in between. Maybe we can collaborate.

**Comment by coder12git** [comment]
> @coder12git I'm working on it since yesterday, but im still unable to relate the CORS Error source with the delay in having the login page, and the stateless screen in between. Maybe we can collaborate.

Okay, I am in let's talk on slack

**Comment by Shiva953** [comment]
@coder12git I've created the thread, please check.

**Comment by coder12git** [comment]
@theBeginner86 can you please assign this issue to me and @Shiva953 , as we both are collaborating on this.

**Comment by coder12git** [comment]
@theBeginner86 can i know how we are checking whether the user is authenticated or not before showing the dashboard component here -
https://github.com/meshery/meshery/blob/8941ee69ecbf874b8dff5c53948e736ea1dde07a/ui/pages/index.js#L16-L24

**Comment by theBeginner86** [comment]
> @theBeginner86 can i know how we are checking whether the user is authenticated or not before showing the dashboard component here -
> 
> https://github.com/meshery/meshery/blob/8941ee69ecbf874b8dff5c53948e736ea1dde07a/ui/pages/index.js#L16-L24

@coder12git 
We use this middleware to determine the user authentication status. https://github.com/meshery/meshery/blob/master/server/handlers/middlewares.go#L60
It introspected the user token and determines if it's still valid or not. If not valid then it redirects to login screen.

**Comment by theBeginner86** [comment]
> @theBeginner86 Have you observed this behaviour (CORS issue) only in the playground or do you also face this when running a local meshery server pointing to the remote provider?

Yes, we do see this issue locally too when running the playground build. You can try with `make server-playground` (change the remote provider URL to local provider)

**Comment by sandramsc** [comment]
This issue has been open for some time with no recent activity, unassigning to open it up for new contributors to give it a go.

**Comment by matrixkavi** [comment]
This issue no longer exist in playground. Hence closing it
